src/devices/src/virtio/block/device.rs

```diff
@@ -121,6 +147,31 @@ impl DiskProperties {
         }
         config
     }
+
+    pub fn cache_type(&self) -> CacheType {
+        self.cache_type
+    }
+}
+
+impl Drop for DiskProperties {
+    fn drop(&mut self) {
+        match self.cache_type {
+            CacheType::Writeback => {
+                // flush() first to force any cached data out.
+                if self.file.flush().is_err() {
+                    error!("Failed to flush block data on drop.");
+                }
+                // Sync data out to physical media on host.
+                if self.file.sync_all().is_err() {
+                    error!("Failed to sync block data on drop.")
+                }
+                METRICS.block.flush_count.inc();
+            }
+            CacheType::Unsafe => {
+                // This is a noop.
```

Comment:
There is no need to include `file.flush()` because `std::fs::File` is not buffered and that is by design. It works the same way an FD would work in C when using `open`, `read` and `write`.

For buffered IO in C, one would use `fopen`, `fread` and `fwrite`. In Rust, this is achieved by using `BufReader` and `BufWriter`.

This means that all data written with a `file.write()` will reach the cache as long as the file is not opened with `O_DIRECT` or `O_DSYNC` (which is not the case in Firecracker), and this is exactly the reason the Rust `file.flush()` is a noop. You can check this and the non-buffered behavior of `file.flush()` using `strace`.

Moreover, I don't think the argument that the `File` implementation might change supports adding `file.flush()`. If anything, it prohibits it. Right now, `file.flush()` is a noop, which means the behviour is the same as not having it. If it does _anything_ in the future and we add it now in the implementation, it would create another bug.

Reply:
My arguments were also rooted in the fact that for `CacheType::WriteBack`, we do a `file.flush()` before calling f`sync()`. why is that?

Even if this a noop, this just follows the semantic of the `flush` method of the `Write` trait

Reply:
It is just future proofing. I thought about the semantics of `File::flush` possibly changing in the future and I wanted to be on the safe side. The rationale was that it has no performance cost now, as it is a noop, and could save us some trouble in the future in the unlikely event that flush semantics change.

I would agree to removing it, I have no strong preference here. Let me know what you think.